fix --komi auto depreciation warning message

wonderingabout opened the pull request:
@roy7 @Dorus @windo 

the previous one was displaying the warning
everytime, because of the unappropriate use
of .includes

replaced with a simpler and correct code now

tested to work :+1:

wonderingabout commented:
oops one typo, rebasing quick

wonderingabout commented:
actually dont merge, i found a much better way to do this :+1:

wonderingabout commented:
yes it does, but now i want to do it without a "if" pre-check :+1: 

using `Array.from()` and `.forEach`

why i want this will make sense in my next PR :100: @windo

wonderingabout commented:
in the end it will not be needed, but i still wanted to try to code something fancy for fun (wont be included in this PR) :+1: 

```
    if (argv.komi) {
        Array.from(argv.komi).forEach(e => {
            if (e === "auto") {
                console.log("Warning: /--komi auto/ has been renamed to /--komi automatic/\n");
                // this argv may be undefined if not used by bot admin, so we skip a line here, not below
            }
        });
    }
```

current error : 

```
/usr/lib/node_modules/gtp2ogs/config.js:669
});
 ^

SyntaxError: Unexpected token )
```

any idea why it doesnt work @windo ?

wonderingabout commented:
oh nvm it was indeed working !!

the problem was i removed a `}` at the `--maxactivegames` just above depreciation message, now works !! :+1:

windo commented on config.js:
```diff
@@ -286,30 +286,35 @@ exports.updateFromArgv = function() {
     }
 
     if (argv.komi) {
-        if (argv.komi.includes(`auto`)) {
```
Since we have this repetition of ranked/unranked/null for a bunch of arguments (and also blitz, live and corr), you could consider in general using a pattern of defining a function that would do the check to reduce trivial repetition:

```
function warn_for_auto_komi(rankedunranked) {
    let arg = 'komi' + rankedunranked;
    for (let e of argv[arg].split(",")) {
        if (e === "auto") {
            console.log("Warning: /--"+arg+" auto/ has been renamed to /--"+arg+" automatic/\n");
            // this argv may be undefined if not used by bot admin, so we skip a line here, not below
        }
    }
}
```

And then:

```
for (let rankedunranked of ['', 'ranked', 'unranked']) {
    warn_for_auto_komi(rankedunranked);
    // maybe other checks here
}
```

wonderingabout replied:
thats exactly what i was doing currently, it is unfinished but you can have a look below @windo

wonderingabout replied:
we are on the same wavelength it seems looo :+1:

wonderingabout replied:
@windo 

after yesterday's review with you, and @Dorus and @roy7 's long review, i now can understand this code and yes its more efficient

Dorus replied:
Why not `"a,b,c".split(',').includes("auto")`?

wonderingabout replied:
i was about to use an array now @Dorus same as in #152 

good idea indeed :+1:

wonderingabout replied:
just rewrote it entirely, closing this so that we can review again :100:

wonderingabout commented:
@windo 

unfinished, will be in PR 2 : 

```
    // console : greeting //

    console.log("\nYou are using gtp2ogs version 6.0\n- For changelog or latest devel updates, please visit https://github.com/online-go/gtp2ogs/tree/devel\n");

    /* console : DEPRECIATIONS */
    console.log("-----DEPRECIATIONS-----");
    let consoleDepreciationAlert = false;

    function depreciatedArgumentIsUsedNewArgumentToUseInstead(depreciatedArgumentUsed, newArgumentToUseInstead) {
        if (argv.depreciatedArgumentUsed) {
            console.log(`Warning: --${depreciatedArgumentUsed} is no longer supported. Use --${newArgumentToUseInstead} instead.\n Read latest up to date available gtp2ogs arguments in OPTIONS-LIST (devel branch) for details : https://github.com/online-go/gtp2ogs/blob/devel/docs/OPTIONS-LIST.md`);
        }
    }

    depreciatedArgumentIsUsedNewArgumentToUseInstead(botid,username);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(bot,username);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(id,username);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(minrankedhandicap,minhandicapranked);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(minunrankedhandicap,minhandicapunranked);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(maxrankedhandicap,maxhandicapranked);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(maxunrankedhandicap,maxhandicapunranked);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(botid,username);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(maxtotalgames,maxconnectedgames);
    depreciatedArgumentIsUsedNewArgumentToUseInstead(maxactivegames,maxconnectedgamesperuser);

    function depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(depreciatedArgumentUsed, newArgumentToUseInsteadOne, newArgumentToUseInsteadTwo, newArgumentToUseInsteadThree,) {
        if (argv.depreciatedArgumentUsedOne) {
            console.log(`Warning: --${depreciatedArgumentUsed} is no longer supported. Use --${newArgumentToUseInsteadOne} and/or --${newArgumentToUseInsteadTwo} and/or --${newArgumentToUseInsteadThree} instead.\nDefaults are provided for all the new arguments, so you dont need to use them all\nRead latest up to date available gtp2ogs arguments in OPTIONS-LIST (devel branch) for details : https://github.com/online-go/gtp2ogs/blob/devel/docs/OPTIONS-LIST.md`);
        }
    }

    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(minmaintime, minmaintimeblitz, minmaintimelive, minmaintimecorr);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(minmaintimeranked, minmaintimeblitzranked, minmaintimeliveranked, minmaintimecorrranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(minmaintimeunranked, minmaintimeblitzunranked, minmaintimeliveunranked, minmaintimecorrunranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(maxmaintime, maxmaintimeblitz, maxmaintimelive, maxmaintimecorr);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(maxmaintimeranked, maxmaintimeblitzranked, maxmaintimeliveranked, maxmaintimecorrranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(maxmaintimeunranked, maxmaintimeblitzunranked, maxmaintimeliveunranked, maxmaintimecorrunranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(minperiodtime, minperiodtimeblitz, minperiodtimelive, minperiodtimecorr);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(minperiodtimeranked, minperiodtimeblitzranked, minperiodtimeliveranked, minperiodtimecorrranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(minperiodtimeunranked, minperiodtimeblitzunranked, minperiodtimeliveunranked, minperiodtimecorrunranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(maxperiodtime, maxperiodtimeblitz, maxperiodtimelive, maxperiodtimecorr);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(maxperiodtimeranked, maxperiodtimeblitzranked, maxperiodtimeliveranked, maxperiodtimecorrranked);
    depreciatedArgumentIsUsedNewArgumentToUseInsteadUpToThree(maxperiodtimeunranked, maxperiodtimeblitzunranked, maxperiodtimeliveunranked, maxperiodtimecorrunranked);

    function isArgumentArrayEqualToAuto(argumentToTest) {
        // we add a if(argv.argumentToTest) before the "auto" test
        // so that we do the test only if argument is defined (used by bot admin)
        if (argv.argumentToTest) {
            for (let e of argv.argumentToTest.split(",")) {
                if (e === "auto") {
                    console.log(`Warning: /--${argumentToTest} auto/ has been renamed to /--${argumentToTest} automatic/`); 
                }
            }
        }
    }

    isArgumentArrayEqualToAuto(komi);
    isArgumentArrayEqualToAuto(komiranked);
    isArgumentArrayEqualToAuto(komiunranked);

    // final [OK!] check
    if (this.consoleDepreciationIsTrue) {
        console.log("result : [ERRORS]\n"); /*IF there is a warning, we skip a line to make it more pretty*/
    
    } else {
        console.log("result : [OK!]\n")
    }

    /* console : WARNINGS, dont use 3 settings of the same family (general, ranked, unranked) at the same time */
    console.log("-----WARNINGS GENERAL/RANKED/UNRANKED-----");
    this.consoleWarningGeneralRankedUnrankedIsTrue = false;

    function generalArgumentIsUsedWithSpecificAtTheSameTime(generalArgument,rankedArgument,unrankedArgument) {
        if (generalArgument && (rankedArgument || unrankedArgument)) {
            console.log(`Warning: You are using --${generalArgument} in combination with --${rankedArgument} and/or --${unrankedArgument}.\n Use either --${generalArgument} alone, OR --${rankedArgument} with --${unrankedArgument}.\nBut don't use both general and specific ranked/unranked arguments at the same time`);
        }
    }

    generalArgumentIsUsedWithSpecificAtTheSameTime(minrank,minrankranked,minrankunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(maxrank,maxrankranked,maxrankunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(boardsize,boardsizeranked,boardsizeunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(komi,komiranked,komiunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(speed,speedranked,speedunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(timecontrol,timecontrolranked,timecontrolunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(minhandicap,minhandicapranked,minhandicapunranked);
    generalArgumentIsUsedWithSpecificAtTheSameTime(,,);
    generalArgumentIsUsedWithSpecificAtTheSameTime(,,);
    generalArgumentIsUsedWithSpecificAtTheSameTime(,,);
    generalArgumentIsUsedWithSpecificAtTheSameTime(,,);
    generalArgumentIsUsedWithSpecificAtTheSameTime(,,);


    if (argv.minhandicap && (argv.minhandicapranked || argv.minhandicapunranked)) {
        console.log("Warning: You are using --minhandicap in combination with --minhandicapranked and/or --minhandicapunranked.\nUse either --minhandicap alone, OR --minhandicapranked with --minhandicapunranked. \nBut don't use the 3 minhandicap arguments at the same time.");
    }

    if (argv.maxhandicap && (argv.maxhandicapranked || argv.maxhandicapunranked)) {
        console.log("Warning: You are using --maxhandicap in combination with --maxhandicapranked and/or --maxhandicapunranked.\nUse either --maxhandicap alone, OR --maxhandicapranked with --maxhandicapunranked.\nBut don't use the 3 maxhandicap arguments at the same time.");
    }

    if (argv.noautohandicap && (argv.noautohandicapranked || argv.noautohandicapunranked)) {
        console.log("Warning: You are using --noautohandicap in combination with --noautohandicapranked and/or --noautohandicapunranked.\nUse either --noautohandicap alone, OR --noautohandicapranked with --noautohandicapunranked.\nBut don't use the 3 noautohandicap arguments at the same time.");
    }

    if (argv.minperiods && (argv.minperiodsranked || argv.minperiodsunranked)) {
        console.log("Warning: You are using --minperiods in combination with --minperiodsranked and/or --minperiodsunranked.\nUse either --minperiods alone, OR --minperiodsranked with --minperiodsunranked.\nBut don't use the 3 minperiods arguments at the same time.");
    }

    if (argv.maxperiods && (argv.maxperiodsranked || argv.maxperiodsunranked)) {
        console.log("Warning: You are using --maxperiods in combination with --maxperiodsranked and/or --maxperiodsunranked.\nUse either --maxperiods alone, OR --maxperiodsranked with --maxperiodsunranked.\nBut don't use the 3 maxperiods arguments at the same time.");
    }

    if (argv.nopause && (argv.nopauseranked || argv.nopauseunranked)) {
        console.log("Warning: You are using --nopause in combination with --nopauseranked and/or --nopauseunranked. \n Use either --nopause alone, OR --nopauseranked with --nopauseunranked.\nBut don't use the 3 nopause arguments at the same time.");
    }

    if (consoleWarningGeneralRankedUnrankedIsTrue

    console.log("\n"); /*after final warning, we skip a line to make it more pretty*/

    // - warning : avoid infinite games
    if (!argv.nopause && !argv.nopauseranked && !argv.nopauseunranked) {
        console.log("Warning : No nopause setting detected, games are likely to last forever"); // TODO : when --maxpaustime and co gets implemented, replace with "are likely to last for a long time"
    }
    console.log("\n"); /*after last warning, we skip a line to make it more pretty*/

    // end of console messages
```

roy7 commented:
Are the changes you discussed with @windo all set and this is ready to merge now?

wonderingabout commented:
@roy7 @Dorus @windo 

i finished writing a code that fixes this 
i'm currently writing a function to generalize it, as the array part will be useful later in the code : 

```
    if (argv.komi || argv.komiranked || argv.komiunranked) {
        let rankedunranked = [ "", "ranked", "unranked"]; // not to modify
        let toModifyRankedunranked = [ "", "ranked", "unranked"]; // example : -> ["komi", "komiranked", "komiunranked"];
        for (let e of toModifyRankedunranked) {
            e = "komi" + e;
            console.log(e);
            if (argv[e]) { // we add a check here to avoid undefined error if bot admin is not using this argv
            // for example if argv[komiranked]
                    //we need to split the argv value into an array before the includes test
                    if (argv[e].split(",").includes("auto")) { 
                        console.log(`Warning: /--${e} auto/ is no longer supported, use /--${e} automatic/ instead`);
                    }
                    //we need to split the argv value into an array before the includes test
                    if (argv[e].split(",").includes("null")) { 
                        console.log(`Warning: /--${e} null/ is no longer supported, use /--${e} automatic/ instead`);
                    }
            }
        }
    }
```
tested to work

also @Dorus `.includes` alone doesnt work, we need to `.split` first as you already know better than me :+1:

wonderingabout commented:
@roy7 @Dorus @windo 

just finished and enhanced this, read commit message for details :+1: 
ready to merge, tested to work :+1: 

![windowarray](https://user-images.githubusercontent.com/38690718/53494720-ae98bd00-3a9e-11e9-8b1c-ce6cf10236aa.png)

wonderingabout commented:
like 

1
```
x = 5
`function f() {
    x = x*x;
}
f();
```

versus : 
2
```
x = 5
`function f() {
    x = 3;
    x = x*x;
}
f();
```

versus : 

3
```
x = 5
`function f(x) {
    x = 3;
    x = x*x;
}
f(x);
```

versus : 
4
```
x = 5
`function f(x) {
    x = x*x;
}
f(x);
```

can you explain me the difference between these 4 cases, its still not clear for me lol :+1: 

1 and 2 work, i know that now, by accessing defined variables out of a function, but what about 3 and 4

can you explain please lol

roy7 commented:
Example A and B are using global variables which is bad programming practice. Try to avoid globals as much as possible. C and D have declaredtheir only local variable named "x" when you write "function f(x)" and that has nothing to do with the global variable of the same name. Another example why globals are bad. :)

Generally any info needed by a function should be passed to the function when the function is called. Now it gets more complex when you are talking about objects like a Game object, and Games have functions that use the data stored in that Game object (ie: this.id or this.state etc).

wonderingabout commented:
oh ok i see @roy7 

clear lol

so f() {
x = blablah}
 is global

but x = "meow"
f(x) {
x = blablah }
is local !!! so not visible in higher order closures !!!

so we'll have for that second case : 
console.log(x) = "meow";
console.log(f(x)) = "blablah" !!!

i see !!!! :+1: 

ok then i'll rewrite this in a better way !! :1st_place_medal:  big thank you, clear now !!!

windo commented:
Global variables are spooky, but A and B examples also work in local scopes. I think that's somewhat less scary, though maybe I'm being foolish in thinking that.

So @roy7, for the use case we were talking about (A single boolean `deprecatedArgFound` defined at outer scope, possibly several helpers needing to update it when they find something), would it be better to update the boolean explicitly:

```
let deprecatedArgFound = false;
function checkDeprecation(argv, argNew, argOld) {
  if (argv[argOld]) {
    ...
    return true;
  }
  return false;
}

deprecatedArgFound |= checkDeprecation(argv, 'oldkomiarg', 'newkomiarg');
deprecatedArgFound |= checkDeprecation(argv, ...);
...
```

I guess that is the clearest way of doing it and not too verbose?

The alternative we were discussing was updating the boolean value as captured implicitly from parent scope:

```
function checkDeprecation(argv, argNew, argOld) {
  if (argv[argOld]) {
    ...
    deprecatedArgFound = true;
  }
}
```

wonderingabout commented:
this is so fun !!! 
i finally understand global and local variables !!! :+1: :+1: :+1: :+1: :+1: :+1: 

so i decided to make generalArg a local variable (we dont need it outside of the function)
but i prefered to keep familyArray a global variable, storing it for later use, and flushing it at every new use :100: 

just rebased !!!!
result : 

![yeeee](https://user-images.githubusercontent.com/38690718/53513592-e830ee80-3ac5-11e9-8380-03049382848e.png)



**thank you for your help everyone** @roy7 @Dorus @windo !!!
you may want to have a look at this too @anoek , it may be basic, but i'm quite proud of what i did lol, feel free to hint improvements or mistakes i made :+1: 

ready to merge, and this function will help a lot for other parts of the code !!!

roy7 commented on config.js:
```diff
@@ -334,6 +313,33 @@ exports.updateFromArgv = function() {
     deprecatedArgs.forEach(ar => testDeprecated(...ar))
+
+    let familyArray = []; 
+    function familyArrayFromGeneralArg(generalArg) {
```
I wonder if the function should return an array?

wonderingabout replied:
interesting idea, i didnt think of it enough

wonderingabout replied:
done by dorus code :+1:

roy7 commented on config.js:
```diff
@@ -334,6 +313,33 @@ exports.updateFromArgv = function() {
       ]
     deprecatedArgs.forEach(ar => testDeprecated(...ar))
+
+    let familyArray = []; 
```
I wonder if we should

`let familyArray = familyArrayFromGeneralArg("komi");`

down after the function is specified.

wonderingabout replied:
oh indeed

i forgot that we could do that in javascript

but i wonder if the familyArray autoflushes at next use now

roy7 commented on config.js:
```diff
@@ -334,6 +313,33 @@ exports.updateFromArgv = function() {
+            e = generalArg + e; // example : e = "komi" + "ranked";
+            familyArray.push(e);
+        }
+    }
```
Last line of function could

`return arr;`

wonderingabout replied:
oh i see, 

yes indeed we can,
but if we do that we may call the functions more times than needed, so maybe storing the array result in a variable is more ressource efficient ? i really dont know, this is just a question :+1:

wonderingabout replied:
yes, maybe thats a good idea

i'll try it when i come back later today

wonderingabout replied:
done by dorus code :+1:

roy7 commented on config.js:
```diff
@@ -334,6 +313,33 @@ exports.updateFromArgv = function() {
+        }
+    }
+
+    familyArrayFromGeneralArg("komi");
```
And then this would be deleted.

wonderingabout replied:
yes, if the function is a defined variable, we would not need this line anymore

but for the next times the function is called,
i think we should not use `let` because familyArray was already defined earlier right ?

for example : 

```let familyArray = familyArrayFromGeneralArg("komi");

// then later in the code, we dont use let again

familyArray = familyArrayFromGeneralArg("boardsize");
```

windo replied:
That's right.

A step further would be to not have an intermediate name for the generated array at all:

```
for (let e of familyArrayFromGeneralArg("komi")) {
    ...
}
for (let e of familyArrayFromGeneralArg("boardsize")) {
    ...
}
```

roy7 commented:
Added some comments, the function could return an array so no use of variables outside of the function is required. Would be cleaner to use in multiple locations too?

wonderingabout commented:
> Would be cleaner to use in multiple locations too?

@roy7 
yes definitely, this function code will be useful to clean many parts of the code

after we generate an array of an argument to test argumentgeneral,argumentranked,argumentunranked

then we can create a loop on top of that and greatly reduce code size and repetitions :+1: 
the komi warning will be removed eventually, but this function will be very useful in the future

wonderingabout commented:
when i come back later today, i'll look at this again

(busy with PhoenixGo tensorrt 5.0 too lol)

roy7 commented:
Correct, you'd only use 'let' the first time you create the variable.

Might be less readable but we could even do

`for (let e of familyArrayFromGeneralArg("boardsize")) {`

And not even store it.

wonderingabout commented:
no its a good idea actually

```
for (let e of familyArrayFromGeneralArg("boardsize")) {
```
would repeat the test we want for every element : 
- boardsize
- boardsizeranked
- boardsizeunranked

looks very good :+1: @roy7

wonderingabout commented:
i'm looking forwad to using a similar array generator for UHMAEATs

there are 18 combinations in total 

min/max
maintime
blitz/live/corr
general/ranked/unranked

and same for periodtimes !!

Dorus commented:
You could just do

```
function familyArrayFromGeneralArg(generalArg) {
    return ["", "unranked", "ranked" ].map(e => generalArg + e)
}
```

Right?

wonderingabout commented:
yes, i should add your code @Dorus and credit you for it :1st_place_medal:

wonderingabout commented:
for my souvenirs, this was the old code (currently committing the new code) : 

```

    let familyArray = []; 
    function familyArrayFromGeneralArg(generalArg) {
        familyArray = []; // first, we reinitialize it to empty at every start
                  // then, for example, will be ["komi", "komiranked", "komiunranked"];
        let arr = [ "", "ranked", "unranked"]; // our source to inject on "komi" for example
        for (let e of arr) {
            e = generalArg + e; // example : e = "komi" + "ranked";
            familyArray.push(e);
        }
    }

    familyArrayFromGeneralArg("komi");

    for (let e of familyArray) {
        if (argv[e]) { // we add a check here to avoid undefined error if bot admin is not using this argv
        // for example if argv[komiranked]
            if (argv[e].split(",").includes("auto")) {
            // we need to split the argv value into an array before the includes test
                console.log(`Warning: /--${e} auto/ is no longer supported, use /--${e} automatic/ instead`);
            }
            if (argv[e].split(",").includes("null")) {
            // we need to split the argv value into an array before the includes test
                console.log(`Warning: /--${e} null/ is no longer supported, use /--${e} automatic/ instead`);
            }
        }
    }
```

wonderingabout commented:
@roy7 @Dorus 

added dorus function which was much more efficient

didnt test it but considering my old code was already working, 
this should work too i hope :+1:
